esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp

```diff
@@ -1491,40 +1636,79 @@ bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDL
+    if (cmp1 != 0)
+        return cmp1;
+    if (itm1->getPort() < itm2->getPort())
+        return -1;
+    else if (itm1->getPort() > itm2->getPort())
+        return 1;
+    else
+        return 0;
+}
 
-    Owned<IPropertyTree> esdlBindings = m_esdlStore->getBindings();
-    if(esdlBindings.get() == nullptr)
-        return false;
-/*Rodrigo
+bool CWsESDLConfigEx::onListESDLBindings(IEspContext &context, IEspListESDLBindingsRequest&req, IEspListESDLBindingsResponse &resp)
+{
     if (m_isDetachedFromDali)
         throw MakeStringException(-1, "Cannot list ESDL Bindings. ESP is currently detached from DALI.");
 
-    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_DESDL);
-    if (!conn)
-       throw MakeStringException(-1, "Unable to connect to ESDL Service definition information in dali '%s'", ESDL_DEFS_ROOT_PATH);
-
-    conn->close(false); //release lock right away
-
-    IPropertyTree * esdlBindings = conn->queryRoot();
-    if (!esdlBindings)
-       throw MakeStringException(-1, "Unable to open ESDL Service definition information in dali '%s'", ESDL_DEFS_ROOT_PATH);
-
-rodrigo*/
+    Owned<IPropertyTree> esdlBindings = m_esdlStore->getBindings();
+    if (esdlBindings.get() == nullptr)
+        return false;
     Owned<IPropertyTreeIterator> iter = esdlBindings->getElements("Binding");
-
+    double ver = context.getClientVersion();
     IArrayOf<IEspESDLBinding> list;
     ForEach(*iter)
     {
         IPropertyTree &item = iter->query();
         Owned<IEspESDLBinding> esdlbinding = createESDLBinding("","");
         esdlbinding->setId(item.queryProp("@id"));
-        esdlbinding->setEspBinding(item.queryProp("@espbinding"));
         esdlbinding->setEspProcess(item.queryProp("@espprocess"));
+        const char* portstr = item.queryProp("@port");
+        if (portstr && *portstr)
+            esdlbinding->setPort(atoi(portstr));
+        else
+            esdlbinding->setPort(0);
+        esdlbinding->setEspBinding(item.queryProp("@espbinding"));
         list.append(*esdlbinding.getClear());
     }
+    list.sort(bindingCompareFunc);
+    if (ver >= 1.4)
+    {
+        IArrayOf<IConstEspProcessStruct>& processes = resp.getEspProcesses();
+        IConstESDLBinding* lastBinding = nullptr;
+        Owned<IEspEspProcessStruct> currentProcess = nullptr;
+        Owned<IEspEspPortStruct> currentPort = nullptr;
+        for (int i = 0; i < list.length(); i++)
+        {
+            IConstESDLBinding* binding = dynamic_cast<IConstESDLBinding*>(&list.item(i));
+            bool processCreated = false;
+            if (!lastBinding || (strcmp(lastBinding->getEspProcess(), binding->getEspProcess()) != 0))
+            {
+                currentProcess.setown(createEspProcessStruct());
+                currentProcess->setName(binding->getEspProcess());
+                processes.append(*dynamic_cast<IConstEspProcessStruct*>(currentProcess.getLink()));
```

Comment:
This cast shouldn't be needed (IConstEspProcessStruct is a base class of IEspEspProcessStruct), and definitely should not be a dynamic cast.